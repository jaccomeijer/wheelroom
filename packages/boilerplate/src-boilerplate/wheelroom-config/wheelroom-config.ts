import { WheelroomConfig } from '@wheelroom/wheelroom'
import { boilerplateTemplates } from './template-sets/boilerplate/boilerplate-templates'
import { graphqlTemplates } from './template-sets/graphql/graphql-templates'
import { contentSets } from './content-sets/content-sets'
import { replaceSets } from './replace-sets'

// The files are in a temp local folder, because of the ts compiler
import { coreModels } from './will-move-a-module/models/core-models'
import { tableModels } from './will-move-a-module/models/table-models'
import { blogModels } from './will-move-a-module/models/blog-models'
import { fieldDefautls } from './will-move-a-module/field-defaults'
import { commonFields } from './will-move-a-module/common-fields'
import { contentfulFieldDefinitions } from './will-move-a-module/contentful-field-definitions'

export const config: WheelroomConfig = {
  commonFields: commonFields,
  components: { ...coreModels, ...blogModels, ...tableModels },
  fieldDefaults: fieldDefautls,
  plugins: [
    {
      options: {
        contentSets,
        fieldDefinitions: contentfulFieldDefinitions,
        replaceSets,
      },
      resolve: '@wheelroom/wheelroom-plugin-contentful',
    },
    {
      options: {
        templateSets: boilerplateTemplates,
      },
      resolve: '@wheelroom/wheelroom-plugin-boilerplate',
    },
    {
      options: {
        templateSets: graphqlTemplates,
      },
      resolve: '@wheelroom/wheelroom-plugin-graphql',
    },
  ],
}
