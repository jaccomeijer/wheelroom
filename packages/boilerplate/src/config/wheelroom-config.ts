import { WheelroomConfig } from '@wheelroom/wheelroom'
import { configCommonFields } from './wheelroom/common-fields'
import { models } from './wheelroom/models'
import { configFieldDefaults } from './wheelroom/field-defaults'
import { configPluginOptions } from './wheelroom/plugin-options'

export const config: WheelroomConfig = {
  commonFields: configCommonFields,
  components: models,
  fieldDefaults: configFieldDefaults,
  plugins: configPluginOptions,
}
