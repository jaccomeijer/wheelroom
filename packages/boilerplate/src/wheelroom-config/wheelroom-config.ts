import { WheelroomConfig } from '@wheelroom/wheelroom'
import { boilerplateTemplates } from './template-sets/boilerplate/boilerplate-templates'
import { graphqlTemplates } from './template-sets/graphql/graphql-templates'
import { contentSets } from './content-sets/content-sets'
import { replaceSets } from './replace-sets'

// The files are in a temp local folder, because of the ts compiler
import { coreModels } from '@wheelroom/core'
import { tableModels } from '@wheelroom/wheel-table'
import { blogModels } from '@wheelroom/wheel-blog'
import { textModels } from '@wheelroom/wheel-text'
import { topicModels } from '@wheelroom/wheel-topic'
import { fieldDefautls } from '@wheelroom/core'
import { commonFields } from '@wheelroom/core'
import { contentfulFieldDefinitions } from '@wheelroom/core'

export const config: WheelroomConfig = {
  commonFields: commonFields,
  components: {
    ...coreModels,
    ...blogModels,
    ...tableModels,
    ...topicModels,
    ...textModels,
  },
  fieldDefaults: fieldDefautls,
  plugins: [
    {
      options: {
        contentSets,
        fieldDefinitions: contentfulFieldDefinitions,
        replaceSets,
      },
      resolve: '@wheelroom/wheelroom-plugin-contentful',
    },
    {
      options: {
        templateSets: boilerplateTemplates,
      },
      resolve: '@wheelroom/wheelroom-plugin-boilerplate',
    },
    {
      options: {
        templateSets: graphqlTemplates,
      },
      resolve: '@wheelroom/wheelroom-plugin-graphql',
    },
  ],
}
