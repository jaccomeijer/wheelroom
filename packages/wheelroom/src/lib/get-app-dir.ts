export const getAppDir = () => {
  return process.cwd()
}
