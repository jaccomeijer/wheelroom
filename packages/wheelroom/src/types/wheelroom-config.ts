import { Plugin } from './plugins'
import { WheelroomComponents } from './wheelroom-components'
import { WheelroomFields } from './wheelroom-fields'
import { CommonField, FieldType, FieldTypeName } from './wheelroom-fields'

export type FieldTypeDefaults = Record<FieldTypeName, FieldType>

export interface FieldDefaults {
  fieldTypeDefaults: FieldTypeDefaults
  commonFieldDefaults: CommonField
}

export interface WheelroomConfig {
  commonFields: WheelroomFields
  components: WheelroomComponents
  fieldDefaults: FieldDefaults
  plugins: Plugin[]
}
