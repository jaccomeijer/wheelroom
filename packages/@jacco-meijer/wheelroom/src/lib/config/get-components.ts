import { ReplaceVars } from '../../types/parser'
import {
  WheelroomComponent,
  WheelroomComponents,
} from '../../types/wheelroom-components'
import { WheelroomConfig } from '../../types/wheelroom-config'
import { FieldType } from '../../types/wheelroom-fields'
import { createParser } from '../parser/create-parser'
import { replaceFunctions } from '../parser/replace-functions'

export const getComponents = async (wheelroomConfig: WheelroomConfig) => {
  if (
    !('components' in wheelroomConfig) ||
    typeof wheelroomConfig.components !== 'object'
  ) {
    console.log('error: components object not found in wheelroom config')
  }
  if (Object.entries(wheelroomConfig.components).length < 1) {
    console.log('error: no components found in wheelroom config')
  }
  const finalComponents = {} as WheelroomComponents
  // Create a parser
  const parser = createParser({
    components: wheelroomConfig.components,
  } as ReplaceVars)
  parser.addReplaceFunctions(replaceFunctions)

  Object.entries(wheelroomConfig.components).forEach(
    // Iterate over all components
    ([componentName, component]: [string, WheelroomComponent]) => {
      parser.updateVars({ component, componentName } as ReplaceVars)

      // Create a working copy for the component
      const workComponent = {
        fields: Object.assign({}, component.fields),
        modelVersion: component.modelVersion,
        settings: component.settings,
      } as WheelroomComponent

      // Merge in common fields
      Object.assign(workComponent.fields, wheelroomConfig.commonFields || {})

      Object.entries(workComponent.fields).forEach(
        // For each component, iterate over all fields
        ([fieldName, field]: [string, FieldType]) => {
          /**
           * Create a working copy of the field with default fields. System
           * fields are only used for building settings queries. We only need the
           * default type field for these.
           */
          let workField = {} as FieldType
          if (!field.system) {
            workField = Object.assign(
              {},
              wheelroomConfig.fieldDefaults.commonFieldDefaults || {}
            ) as FieldType
          }
          // Copy field type specific defaults
          Object.assign(
            workField,
            wheelroomConfig.fieldDefaults.fieldTypeDefaults[field.type]
          )
          // Copy field attributes to working copy
          Object.assign(workField, field)
          // Create an object for saving parse results
          const parseResults = {} as {
            [attr: string]: string | string[]
          }
          Object.entries(workField).forEach(
            // Parse variables of all strings
            ([key, value]: [string, string]) => {
              // Only parse strings and arrays of string
              if (typeof value !== 'string' && !Array.isArray(value)) {
                return
              }
              parser.updateVars({
                field,
                fieldName,
                fieldType: field.type,
              } as ReplaceVars)
              parseResults[key] = parser.parse(value)
            }
          )
          // Merge parse results to field working copy
          Object.assign(workField, parseResults)
          // Update component working copy with field working copy
          workComponent.fields[fieldName] = workField
        }
      )
      // Save working component copy to our object with final components
      finalComponents[componentName] = workComponent
    }
  )
  return finalComponents
}
