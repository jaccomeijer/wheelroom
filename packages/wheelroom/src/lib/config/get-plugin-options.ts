import { WheelroomConfig } from '../../types/wheelroom-config'

export const getPluginOptions = (config: WheelroomConfig) => {
  return config.plugins.reduce((result: any, plugin) => {
    result[plugin.resolve] = plugin.options
    return result
  }, {})
}
