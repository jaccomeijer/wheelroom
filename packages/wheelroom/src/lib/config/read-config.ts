import { WheelroomConfig } from '../../types/wheelroom-config'

/**
 * Read wheelroom-config.js from a given path
 */
export const readConfig = async (configPath: string, customTarget?: string) => {
  let config
  const target = customTarget
    ? `${configPath}/${customTarget}`
    : `${configPath}/wheelroom-config.js`
  try {
    config = await import(target)
  } catch (error) {
    if (error.code === 'MODULE_NOT_FOUND') {
      console.log(`Could not find module: ${target}`)
    } else {
      // In case of a parser error, print full error
      console.log(error)
    }
    return
  }
  return config.config as WheelroomConfig
}
