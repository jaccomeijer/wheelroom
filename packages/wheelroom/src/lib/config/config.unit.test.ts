import { config } from './fixtures/wheelroom-config'
import { getAppDir } from '../get-app-dir'
import { getComponents } from './get-components'
import { getPluginOptions } from './get-plugin-options'
import { readConfig } from './read-config'

/**
 *  This test needs the compiled fixture in dist/lib/config/fixtures. To compile run
 *
 *  npm run build:compile
 */

describe('Readig config should provide', () => {
  test('the whole config from COMPILED! dist/lib/config/fixtures/wheelroom-config.js', async () => {
    const configRead = await readConfig(
      getAppDir(),
      'dist/lib/config/fixtures/wheelroom-config.js'
    )
    expect(configRead).toStrictEqual(config)
  })

  test('the configured components for en-US locale', async () => {
    const components = await getComponents(config, 'en-US')
    expect(components).toMatchSnapshot()
  })

  test('the configured components for nl locale', async () => {
    const components = await getComponents(config, 'nl')
    expect(components).toMatchSnapshot()
  })

  test('the configured plugin options', async () => {
    const options = getPluginOptions(config)
    expect(options).toMatchSnapshot()
  })
})
