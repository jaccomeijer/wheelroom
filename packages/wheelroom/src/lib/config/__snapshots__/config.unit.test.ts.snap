// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Readig config should provide the configured components 1`] = `
Object {
  "allFieldTypes": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "dateField": Object {
        "initialContent": "2019-06-03T00:00+01:00",
        "localized": false,
        "name": "Date field",
        "required": false,
        "type": "date",
      },
      "dropdownField": Object {
        "initialContent": "item A",
        "items": Array [
          "item A",
          "item B",
          "item C",
        ],
        "localized": false,
        "name": "Dropdown field",
        "required": false,
        "type": "dropdown",
      },
      "imageField": Object {
        "initialContent": true,
        "localized": false,
        "name": "Image field",
        "required": false,
        "type": "media",
      },
      "longTextField": Object {
        "initialContent": "This is a long text field",
        "localized": false,
        "name": "Long text field",
        "required": false,
        "type": "longText",
      },
      "multipleComponentsField": Object {
        "allowedComponents": Array [
          "page",
          "firstSection",
          "secondSection",
        ],
        "initialContent": Array [
          "firstSection",
          "secondSection",
        ],
        "localized": false,
        "name": "Multiple components field",
        "required": false,
        "type": "multipleComponents",
      },
      "numberField": Object {
        "initialContent": 5,
        "localized": false,
        "name": "Number field",
        "required": false,
        "type": "number",
      },
      "richTextField": Object {
        "initialContent": "This is a rich text field",
        "localized": false,
        "name": "Rich text field",
        "required": false,
        "type": "richText",
      },
      "shortTextField": Object {
        "initialContent": "This is a short text field",
        "localized": false,
        "maxLength": 155,
        "name": "Short text field",
        "required": false,
        "type": "shortText",
      },
      "singleComponentField": Object {
        "allowedComponents": Array [
          "pageModel",
        ],
        "initialContent": "pageModel",
        "localized": false,
        "name": "Single component field",
        "required": false,
        "type": "singleComponent",
      },
      "tagsField": Object {
        "initialContent": Array [
          "tag A",
          "tag B",
        ],
        "localized": false,
        "name": "Tags field",
        "required": false,
        "type": "tags",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "All field types generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asQuery": "subPage",
      "queryLimit": 10,
    },
  },
  "firstSection": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "defaultField": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Default field",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "First section generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asPageSection": true,
    },
  },
  "globalModel": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "heading": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Heading",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "Global model generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asQuery": "global",
      "queryLimit": 10,
    },
  },
  "noFragment": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "heading": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Heading",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "No fragment generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {},
  },
  "pageModel": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "heading": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Heading",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "Page model generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asQuery": "page",
      "queryLimit": 10,
    },
  },
  "secondSection": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "heading": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Heading",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "Second section generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asPageSection": true,
    },
  },
  "subPageModel": Object {
    "fields": Object {
      "__typename": Object {
        "initialContent": "Initial content for short text field",
        "system": true,
        "type": "shortText",
      },
      "heading": Object {
        "initialContent": "Initial content for short text field",
        "localized": false,
        "name": "Heading",
        "required": false,
        "type": "shortText",
      },
      "title": Object {
        "helpText": "Never displayed, only used for listing within Contentful",
        "initialContent": "Sub page model generated from boilerplate",
        "localized": false,
        "name": "Title",
        "required": true,
        "type": "shortText",
      },
    },
    "modelVersion": "1.0.0",
    "settings": Object {
      "asFragment": true,
      "asQuery": "subPage",
      "queryLimit": 10,
    },
  },
}
`;

exports[`Readig config should provide the configured plugin options 1`] = `
Object {
  "another-module-name": Object {
    "anotherTestOption": "passed-to-plugin",
    "testArray": Array [
      "passed",
      "to",
      "plugin",
    ],
  },
  "module-name": Object {
    "testOptione": "to-pass",
  },
}
`;
